Added segment about string types and their corresponding member functions (#87)
onrirr/c3-web: src/content/docs/Language Common/strings.md

---
title: Strings
description: Strings
sidebar:
    order: 62
---
In C3, multiple string types are available, each suited to different use cases.

> #### String

Reviewer: ```suggestion
### String
```
H3 is linkable

```c3

distinct String = inline char[];

```
\

> Strings are usually the typical type to use, they can be sliced , compared etc ... \

Reviewer: Strings are the typical string type to use, they can be sliced, compared, split etc, the [string methods](#string-member-functions) are outlined below.
It is possible to access the length of a `String` instance through the  `.len`  member, eg
```c3
String my_string = "example";
int length = my_string.len
```

It is possible to access the length of a `String` instance through the  ` .len  `  operator.

> #### ZString

Reviewer: ```suggestion
### ZString
```
so we can link, H3 is linkable

```c3

distinct ZString = inline char*;
```

\

> A ZString is a distinct type for null-terminated C-style strings ` char* `. Converting a ` char* ` to a ` ZString ` requires an explicit cast to ensure the safety and `\0` termination of the string data.

Reviewer: Perhaps we can add the use-case of this string type:

```suggestion
ZString is used when working with C code, which expects null-terminated C-style strings of type `char*`. it is a distinct type so converting to a `ZString` requires an explicit cast. This helps to remind the user to check there is appropriate `\0` termination of the string data. 

The [ZString methods](#zstring-member-functions) are outlined below.
```

> >  **⚠️**  Converting a `String` to a `ZString` should be done carefully to ensure \0 termination of the underlying `char*`.

Reviewer: ```suggestion
:::caution  
Ensure the terminal `\0` when converting from `String` to `ZString`.
:::
```

#### WString

```c3

distinct WString = inline Char16*;
```

\
The `WString` type is similar to `ZString` but uses `Char16*`, typically for UTF-16 encoded strings. This type is useful for applications where 16-bit character encoding is required.

#### DString

```c3

distinct DString (OutStream) = void*;
```

\
`DString` is a dynamic string builder that supports various string operations at runtime, allowing for flexible manipulation without the need for manual memory allocation.

## Member functions:

> #### String :

Reviewer: So we can link directly to this heading (H3 is linkable, H4 is not)
```suggestion
### String Member Functions
```

fn ushort! String.to_ushort(s, int base = 10) => s.to_integer(ushort, base);
```
```c3 implementation
fn char! String.to_uchar(s, int base = 10) => s.to_integer(char, base);
```
```c3 implementation
fn double! String.to_double(s) => s.to_real(double);
```
```c3 implementation
fn float! String.to_float(s) => s.to_real(float);
```
```c3 implementation
fn String String.new_ascii_to_upper(s, Allocator allocator = allocator::heap());
```

```c3 implementation
fn Char16[]! String.to_new_utf16(s, Allocator allocator = allocator::heap());
```

```c3 implementation
fn Char16[]! String.to_temp_utf16(s);
```

```c3 implementation
fn Char32[]! String.to_utf32(s, Allocator allocator);
```

```c3 implementation
fn Char32[]! String.to_new_utf32(s) => s.to_utf32(allocator::heap()) @inline;
```

```c3 implementation
fn Char32[]! String.to_temp_utf32(s) => s.to_utf32(allocator::temp()) @inline;
```

```c3 implementation
fn WString! String.to_wstring(s, Allocator allocator);
```

```c3
fn WString! String.to_new_wstring(s) => s.to_wstring(allocator::heap());
```
```c3
fn WString! String.to_temp_wstring(s) => s.to_wstring(allocator::temp());
```
```c3
fn StringIterator String.iterator(s);
```

> #### ZString :

Reviewer: So we can link directly to this heading (H3 is linkable, H4 is not)
```suggestion
### ZString Member Functions
```
